seed/tests/test_derived_columns.py

```diff
@@ -0,0 +1,626 @@
+
+        models = self._derived_column_for_property_factory(expression, column_parameters)
+        derived_column = models['derived_column']
+        property_state = models['property_state']
+
+        # verify the property state is missing the expected column
+        self.assertTrue('a' not in property_state.extra_data)
+
+        expected_parameter_values = {
+            'a': None,
+            'b': 1
+        }
+
+        # -- Act
+        result = derived_column.get_parameter_values(property_state)
+
+        # -- Assert
+        self.assertDictEqual(expected_parameter_values, result)
+
+    def test_derived_column_evaluate_is_successful_when_property_state_has_valid_values(self):
+        # -- Setup
+        expression = '$a + $b'
+        column_parameters = {
+            'a': {
+                'source_column': self.col_factory('foo', is_extra_data=True),
+                'value': 1,
+            },
+            'b': {
+                'source_column': self.numeric_core_columns[0],
+                'value': 1,
+            }
+        }
+
+        models = self._derived_column_for_property_factory(expression, column_parameters)
+        derived_column = models['derived_column']
+        property_state = models['property_state']
+
+        # -- Act
+        result = derived_column.evaluate(property_state)
+
+        # -- Assert
+        self.assertEqual(2, result)
+
+    def test_derived_column_evaluate_returns_none_when_missing_parameters(self):
+        # -- Setup
+        # put a parameter, x, which won't exist
+        expression = '$a + $x'
+        column_parameters = {
+            'a': {
+                'source_column': self.col_factory('foo', is_extra_data=True),
+                'value': 1,
+            },
+            'b': {
+                'source_column': self.numeric_core_columns[0],
+                'value': 1,
+            }
+        }
+
+        models = self._derived_column_for_property_factory(expression, column_parameters)
+        derived_column = models['derived_column']
+        property_state = models['property_state']
+
+        # -- Act
+        result = derived_column.evaluate(property_state)
+
+        # -- Assert
+        self.assertIsNone(result)
+
+    def test_derived_column_evaluate_returns_none_when_parameter_is_nonnumeric(self):
+        # -- Setup
+        expression = '$a + 1'
+        column_parameters = {
+            'a': {
+                'source_column': self.col_factory('foo', is_extra_data=True),
+                'value': 'HELLO!!!!!',
+            }
+        }
+
+        models = self._derived_column_for_property_factory(expression, column_parameters)
+        derived_column = models['derived_column']
+        property_state = models['property_state']
+
+        # -- Act
+        result = derived_column.evaluate(property_state)
+
+        # -- Assert
+        self.assertIsNone(result)
+
+    def test_derived_column_evaluate_falls_back_to_passed_params_if_not_in_inventory(self):
```

Comment:
What's the use case for passing in "fallback" params? It seems related to not having an inventoryState object to base values off, but it isn't immediately clear to me when that might happen. 

I have a feeling I'm missing/forgetting something notable 😅

Reply:
Yeah I think I just added that without considering it too much. I think "falling back" onto a params dict is unlikely. But passing only a dict is somewhat likely. I just didn't want to make the arguments mutually exclusive, and had to decide a precedence.